pkg/operator/staticpod/controller/prune/prune_controller.go

```diff
@@ -92,85 +87,77 @@ func getRevisionLimits(operatorSpec *operatorv1.StaticPodOperatorSpec) (int32, i
+// - don't prune a node's CurrentRevision and the spec.succeededRevisionLimit - 1 revisions before it.
+// - don't prune a node's TargetRevision and the spec.failedRevisionLimit - 1 revisions before it.
+// - don't prune a node's LastFailedRevision and the spec.failedRevisionLimit - 1 revisions before it.
+func (c *PruneController) revisionsToKeep(status *operatorv1.StaticPodOperatorStatus, failedLimit, succeededLimit int) (all bool, keep sets.Int32) {
+	// find oldest where we are sure it cannot fail anymore (i.e. = currentRevision
+	var oldestSucceeded int32 = maxInt32
+	for _, ns := range status.NodeStatuses {
+		if ns.CurrentRevision < oldestSucceeded {
+			oldestSucceeded = ns.CurrentRevision
 		}
+	}
+	if oldestSucceeded < status.LatestAvailableRevision && failedLimit == -1 {
+		return true, nil
+	}
+	if succeededLimit == -1 {
+		return true, nil
+	}
 
-		if revision, ok := configMap.Data["revision"]; ok {
-			revisionNumber, err := strconv.Atoi(revision)
-			if err != nil {
-				return []int{}, err
-			}
-			switch configMap.Data["status"] {
-			case string(corev1.PodSucceeded):
-				succeededRevisions = append(succeededRevisions, revisionNumber)
-			case string(corev1.PodFailed):
-				failedRevisions = append(failedRevisions, revisionNumber)
-			case StatusAbandoned:
-				abandonedRevisions = append(abandonedRevisions, revisionNumber)
-			case StatusInProgress:
-				// we always protect inprogress
-				inProgressRevisions = append(inProgressRevisions, revisionNumber)
-
-			default:
-				// protect things you don't understand
-				unknownStatusRevisions = append(unknownStatusRevisions, revisionNumber)
-				recorder.Event("UnknownRevisionStatus", fmt.Sprintf("unknown status for revision %d: %v", revisionNumber, configMap.Data["status"]))
-			}
+	keep = sets.Int32{}
+	if oldestSucceeded < status.LatestAvailableRevision {
+		keep.Insert(int32RangeBelowOrEqual(status.LatestAvailableRevision, maxLimit(failedLimit, succeededLimit))...) // max because we don't know about failure or success
+	} // otherwise all nodes are on LatestAvailableRevision already. Then there is no fail potential.
+
+	for _, ns := range status.NodeStatuses {
+		if ns.CurrentRevision > 0 {
+			keep.Insert(int32RangeBelowOrEqual(ns.CurrentRevision, succeededLimit)...)
+		}
+		if ns.TargetRevision > 0 {
+			keep.Insert(int32RangeBelowOrEqual(ns.TargetRevision, maxLimit(failedLimit, succeededLimit))...) // max because we don't know about failure or success
+		}
+		if ns.LastFailedRevision > 0 {
+			keep.Insert(int32RangeBelowOrEqual(ns.LastFailedRevision, failedLimit)...)
 		}
 	}
 
-	// Return early if nothing to prune
-	if len(succeededRevisions)+len(failedRevisions)+len(abandonedRevisions) == 0 {
-		klog.V(2).Info("no revision IDs currently eligible to prune")
-		return []int{}, nil
+	if keep.Len() > 0 && keep.List()[0] == 1 && keep.List()[keep.Len()-1] == status.LatestAvailableRevision {
```

Comment:
mhm, does it mean that on a fresh cluster we can potentially keep all revisions?

Reply:
we better do. Fresh = bootstrapping. I don't want to prune anything at that point.